### PR TITLE
feat: add targer argument to setParams

### PR DESCRIPTION
**Pull request by osdnk**

### CODE REVIEW

File: src/BaseActions.tsx [hunk 1]
```diff
@@ -13,24 +13,28 @@ export type Action =
...
-    (target.hasOwnProperty('key') && target.hasOwnProperty('name')) ||
-    (!target.hasOwnProperty('key') && !target.hasOwnProperty('name'))
-  ) {
-    throw new Error(
-      'While calling navigate you need to specify either name or key'
-    );
-  }
   if (typeof target === 'string') {
     return { type: 'NAVIGATE', payload: { name: target, params } };
   } else {
```
Comment (osdnk) [inline src/BaseActions.tsx:29]: It was a mistake before. Sorry!

File: src/BaseActions.tsx [hunk 2]
```diff
@@ -42,3 +46,20 @@ export function replace(name: string, params?: object): Action {
 export function reset(state: PartialState & { key?: string }): Action {
   return { type: 'RESET', payload: state };
 }
+
+export function setParams(
+  params: object,
+  target: { name?: string; key?: string }
```
Comment (satya164) [inline src/BaseActions.tsx:52]: I think we should make the signature same as `navigate` etc. I think it's fine if it's required.

File: src/BaseRouter.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import { CommonAction, Router } from './types';
+
+const BaseRouter: Omit<
+  Omit<Omit<Router<CommonAction>, 'getInitialState'>, 'getRehydratedState'>,
+  'getRehydratedState'
```
Comment (satya164) [inline src/BaseRouter.tsx:5]: omitting twice

I think we can remove the annotation. it'll be simpler. If we annotate the individual methods, the type of the object should be inferred.

File: src/BaseRouter.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            r.key === action.payload.key || r.name === action.payload.name
+              ? { ...r, params: { ...r.params, ...action.payload.params } }
+              : r
+          ),
+        };
+      default:
+        return null;
+    }
+  },
+  getStateForRouteNamesChange(state, { routeNames }) {
```
Comment (satya164) [inline src/BaseRouter.tsx:23]: nit: I like new lines between methods :)

File: src/BaseRouter.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  },
+  shouldActionPropagateToChildren(action) {
+    return action.type === 'NAVIGATE';
+  },
+
+  shouldActionChangeFocus(action) {
+    return action.type === 'NAVIGATE';
+  },
+
+  actionCreators: {},
```
Comment (satya164) [inline src/BaseRouter.tsx]: since this isn't used, maybe we don't need it?

File: example/StackNavigator.tsx [hunk 6]
```diff
@@ -53,7 +53,6 @@ export type StackNavigationProp<
    */
   popToTop(): void;
 };
-
```
Comment (satya164) [inline example/StackNavigator.tsx]: nit: keep newline

File: example/StackNavigator.tsx [hunk 7]
```diff
@@ -53,7 +53,6 @@ export type StackNavigationProp<
...
 const StackRouter: Router<CommonAction | Action> = {
```
Comment (satya164) [inline example/StackNavigator.tsx:57]: The router should also merge base router no? Then we can remove the methods those are the same

File: src/BaseRouter.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+      ...state,
+      routeNames,
+      routes: state.routes.filter(route => routeNames.includes(route.name)),
+    };
+  },
+
+  getStateForRouteFocus(state, key) {
+    const index = state.routes.findIndex(r => r.key === key);
+
+    if (index === -1) {
```
Comment (satya164) [inline src/BaseRouter.tsx]:
```suggestion
    if (index === -1 || index === state.index) {
```

Comment (osdnk) [reply src/BaseRouter.tsx]: why?

File: src/BaseRouter.tsx (hunk 3 above)
Comment (osdnk) [reply src/BaseRouter.tsx:5]: but it's really simple, isn't it?

File: src/BaseRouter.tsx (hunk 5 above)
Comment (osdnk) [reply src/BaseRouter.tsx]: I made it optional

File: src/BaseRouter.tsx (hunk 8 above)
Comment (satya164) [reply src/BaseRouter.tsx]: we don't need to update state if that route was already focused

**Review by satya164 — APPROVED** [review]

**Pull request merged**